File: auth/src/utils/meta.rs
use chrono::Utc;
use std::fs;

use crate::types;

pub fn get_meta_factors() -> types::MetaFactors {
    let hash = fs::read_to_string(".meta/HASH")
        .unwrap_or("".to_string())
        .trim_end()
        .to_string();

    let version = fs::read_to_string(".meta/VERSION")
        .unwrap_or("".to_string())
        .trim_end()
        .to_string();

    let start_time = Utc::now();

    types::MetaFactors {
        hash,
        version,
        start_time,
    }
}

pub fn get_meta_from_factors(factors: types::MetaFactors) -> types::Meta {
    let uptime = Utc::now() - factors.start_time;

    types::Meta {
        hash: factors.hash,
        version: factors.version,
        uptime: uptime.num_seconds(),
    }
}

File: realtime/utils/meta.go
package utils

import (
	"os"
	"strings"
	"time"

	"github.com/autobar-dev/services/realtime/types"
)

func GetMetaFromFactors(mf *types.MetaFactors) *types.Meta {
	now := time.Now()
	uptime := int64(now.Sub(mf.StartTime).Seconds())

	return &types.Meta{
		Uptime:  uptime,
		Hash:    mf.Hash,
		Version: mf.Version,
	}
}

func GetMetaFactors() *types.MetaFactors {
	commit_sha := ""
	version := ""

	if commit_sha_bytes, err := os.ReadFile(".meta/HASH"); err == nil {
		commit_sha = strings.TrimSpace(string(commit_sha_bytes))
	}

	if version_bytes, err := os.ReadFile(".meta/VERSION"); err == nil {
		version = strings.TrimSpace(string(version_bytes))
	}

	start_time := time.Now()

	return &types.MetaFactors{
		StartTime: start_time,
		Hash:      commit_sha,
		Version:   version,
	}
}
